### ISSUE TITLE
How to know when a component has finished updating DOM?

### ISSUE THREAD

**Issue by PrinceAli321** [knockout/knockout#1475]
Hi Guys,

I need to reflow my page after some KO components have finished rendering in the DOM to support this Foundation plugin: http://foundation.zurb.com/docs/components/equalizer.html

I know we have afterRender for the template and foreach bindings but i can't find anything similar for the new components functionality.

How can i go about executing a callback when my components are loaded (either individually or after all have been loaded)? Is there something i've missed?

**Comment by W3Max** [comment]
+1 for this issue and for @richotaylor comments about child components

**Comment by jods4** [comment]
I solved this problem using an early build of 3.3 and turning all my components synchronous. Other benefits: better perf and less flashing content.

This does not prevent you from dynamically loading component templates, I just do it myself and pass a string to the component declaration. It's only a different way to organize the codebase (and it also makes me think that including `require` logic inside KO components was a bit of a mistake).

**Comment by W3Max** [comment]
@jods4 Could you share a code example of how you do it? I would really appreciate it!

**Comment by jods4** [comment]
Well, in my project I'm using Typescript with requirejs as AMD loader, but in essence it works like this (Sorry if my js syntax for require is not 100% correct, TS makes this so much easier):

Step 1: be sure to have a KO 3.3 build, otherwise this won't work.

Step 2: define your component.
Key points here: I use requirejs to fetch the template as a string and use the option `synchronous: true`.
```javascript
define(['knockout', 'text!template'], function(ko, template) {
  ko.components.register('my-component', {
    viewModel: MyVM,
    template: template,
    synchronous: true
  });
});
```

Step 3: make sure your component is loaded before any view tries to use it. 
You can do as you like, I do so by adding a fake require dependency on any viewmodel that needs a component in its view.
```javascript
define(['MyComponent'], function() {
  /* Define whatever you want that needs MyComponent to be loaded */
});
```

**Comment by W3Max** [comment]
@jods4 Thanks! I will look into ```synchronous: true```

**Comment by jods4** [comment]
@W3Max Note that as of today 3.3 is officially released! Look at @rniemeyer  blog post for more info: http://www.knockmeout.net/2015/02/knockout-3-3-released.html

**Comment by lzl124631x** [comment]
@jods4 @W3Max I want to note that the first time load still might be aync even if you mark the component with `synchronous: true`.

> If you want to change the policy for a particular component, you can specify synchronous: true on that component’s configuration. Then **it might load asynchronously on first use**, followed by synchronously on all subsequent uses. If you do this, then you need to account for this changeable behavior in any code that waits for components to load. However, if your component can always be loaded and initialized synchronously, then enabling this option will ensure consistently synchronous behavior. This might be important if you’re using a component within a foreach binding and want to use the afterAdd or afterRender options to do post-processing.

**Comment by jods4** [comment]
@lzl124631x This is pretty old stuff but what I did was that I bundled all the templates with my app (using require.js + !text plugin). I required them in the file that created the custom component and passed them as string to KO. -> guaranteed fast sync load.

**Comment by lzl124631x** [comment]
Shaking hands. I'm using require.js, !text plugin and r.js for AMD, bundling and minifying. All the js and html files are bundled into one js file in my case. So, as you've said, this can ensure me it's a sync load? I'm not sure about that. Have you tested it?

(Just out of curiosity, are you still using require.js? I'm using it but found that it's very difficult to set the paths configs when your project getting bigger and bigger. Things got messed up. I tend to make a shift to webpack which seems more popular now. Have you made the shift?)

**Comment by jods4** [comment]
@lzl124631x I went as far as _not letting KO load the template through require_ but passing it directly as a string. 
So my module imports the template into a string as a dependency, then creates the KO component and feeds it the template as a string.
This way I am sure composition is synchronous, even the first time.

Although we are moving away from KO -- which is great but doesn't seem to evolve towards newer web standards -- to Aurelia for newer projects, we still rely on AMD/requirejs at the moment. It is stable and has worked great for us for several years. Webpack, browserify and co. look cool but it seems that there is still a lot of friction and moving parts, which we can't afford here at work. We are considering making the move, but haven't done so... yet. :wink:

**Comment by gabidi** [comment]
Hope this helps 
https://github.com/gabidi/Sample-Durandal-preload-template

**Comment by gabidi** [comment]
It allows you to load HTML after the Dom ready event

**Comment by jacobslusser** [comment]
I've been using RequireJS and Knockout on large production sites for a couple years now and I've found that the technique suggested by @jods4 is the way to go. Knockout should not have put `require` support in the box because it just leads to confusion and uncertainty about when a component is loaded.

I got tired though of constantly having to add a dependency in every view model to `'text!<view>'` to get the view loaded and add the code suggested by @jods4 for registering the component.... 

So I put it all into a RequireJS plugin. The relevant plugin snippet is below, but the complete Gist can be found at: [https://gist.github.com/jacobslusser/2d9210606a8ab64f3a4df7747ee40404](https://gist.github.com/jacobslusser/2d9210606a8ab64f3a4df7747ee40404):

```js
// A simple RequireJS plugin that will load a view and view model based on naming convention.
// Once loaded the view is added to the DOM so it can be used with KO templates and the view
// and view model are registered as a KO component.
define('component', ['jquery', 'knockout'], function ($, ko) {
    return {
        load: function (name, parentRequire, onload, config) {
            parentRequire([name, 'text!' + name + '.html'], function (viewModel, view) {

                // Add the view to the DOM
                var template = '<script type="text/template" id="' + name + '">' + view + '</script>';
                $('body').append(template);

                // Register as a KO component
                ko.components.register(name, {
                    viewModel: viewModel,
                    template: { element: name },
                    synchronous: true
                });

                onload(viewModel);
            }, function (err) {
                onload.error(err);
            });
        }
    };
});
```

NOTE: This plugin assumes that `<view>.html` and `<viewmodel>.js` are in the same folder. If they are not, or are in a subfolder, adjust your plugin accordingly.

Put this in it's own file, or as I do, just in my main RequireJS config file and use it like this:

```js
// The component is required using the 'component!' plugin we created. This will load
// 'page1.html' and 'page1.js' and return 'page1.js' to us.
define(['knockout', 'component!page1'], function (ko, Page1) {
    'use strict';

    function App() {

        var args = { name: 'Jacob' };

        // Example of using the component binding with the automatically registered component
        this.sampleComponent = ko.observable({
            name: 'page1',
            params: args
        });

        // Templates can also be used and are probably preferable because we have easy access
        // to the view model if we want to programmatically call anything on the view model.
        // i.e. sampleTemplate().data.yourFunction()
        this.sampleTemplate = ko.observable({
            name: 'page1',
            data: new Page1(args)
        });
    }
    
    ko.applyBindings(new App());
});
```

When the plugin loads the template it also adds it to the DOM and registers the KO component using the `template: { element: template }` syntax rather than the string `template: template` syntax. This has the added benefit that you can use the `'component!'` plugin to require a template. So it works for both components and templates! See the example above.

This has had a huge productivity boost in our large RequireJS/Knockout applications. The result is that we typically use the `component` binding when we don't need to interact with the view model and the `params` we can pass it through the `component` binding are simple. If, however, we want to have a reference to the view model and interact with it by calling functions on the view model or something else, we use the `template` binding which avoids all the nastiness associated with communicating between parent/child view models and offers support for things like the `afterRender` function.

It's the best of both worlds; it's syntactically easier to follow; and has better performance because the view and view model are required in parallel rather than in serial as it is when you require `'<viewmodel>'` and then '`text!<view>'` (if you're not already bundling your modules).

A more complete example can be found in the afformentioned Gist:  [https://gist.github.com/jacobslusser/2d9210606a8ab64f3a4df7747ee40404](https://gist.github.com/jacobslusser/2d9210606a8ab64f3a4df7747ee40404).

Still lovin' Knockout after all these years...

**Comment by str** [comment]
+1 Isn't this possible yet?

**Comment by brianmhunt** [comment]
knockout/tko (ko 4 candidate) latest `master` branch has this.

More specifically, the `applyBindings` family of functions now return a Promise that resolves when sub-children (including asynchronous ones) are bound.

The API isn't set or documented yet, but the bones have been set up.

**Comment by mbest** [comment]
Fixed with #2319.
